cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkflowProgramRunner.java

```diff
@@ -105,11 +105,21 @@ protected ProgramController launch(Program program, ProgramOptions options,
     // Adds the extra class that Spark runtime needed
     ProgramRuntimeProvider provider = runtimeProviderLoader.get(ProgramType.SPARK);
     if (provider != null) {
-      extraDependencies.add(provider.getClass());
+      try {
+        String sparkAssemblyJarName = SparkUtils.prepareSparkResources(tempDir, localizeResources);
+        // Localize the spark-assembly jar and spark conf zip
+        extraClassPaths.add(sparkAssemblyJarName);
+        extraDependencies.add(provider.getClass());
+      } catch (Exception e) {
+        if (driverMeta.hasSpark) {
+          // If the Workflow actually has spark, we can't ignore this error.
+          throw e;
+        }
+        // Otherwise, this can be ignore.
+        LOG.debug("Spark assembly jar is not present. Not affected Workflow {} since it doesn't use Spark.",
```

Comment:
nit: "Does not affect... "

Also, what do users have to do to be able to use spark in a workflow after this? Do they have to restart CDAP after installing spark, so that cached jars are not used? If so, should we add that to the error message?